Port WebExtensionStorageSQLiteStore to C++

ItsJamie9494 opened the pull request:
#### 433f8948107045bc3b0ad5d57caa56b19e43000c
<pre>
Port WebExtensionStorageSQLiteStore to C++
<a href="https://webkit.org/b/286343">https://webkit.org/b/286343</a>

Reviewed by Timothy Hatcher.

This removes the old Cocoa implementation of WebExtensionStorageSQLiteStore, and creates a new c++ storage store
based off the new C++ WebExtensionSQLiteStore

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp: Renamed from Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm.
(WebKit::WebExtensionContext::isStorageMessageAllowed):
(WebKit::WebExtensionContext::storageGet):
(WebKit::WebExtensionContext::storageGetKeys):
(WebKit::WebExtensionContext::storageGetBytesInUse):
(WebKit::WebExtensionContext::storageSet):
(WebKit::WebExtensionContext::storageRemove):
(WebKit::WebExtensionContext::storageClear):
(WebKit::WebExtensionContext::fireStorageChangedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::localStorageStore):
(WebKit::WebExtensionContext::sessionStorageStore):
(WebKit::WebExtensionContext::syncStorageStore):
(WebKit::WebExtensionContext::storageForType):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::getDataRecords):
(WebKit::WebExtensionController::getDataRecord):
(WebKit::WebExtensionController::removeData):
(WebKit::WebExtensionController::calculateStorageSize):
(WebKit::WebExtensionController::removeStorage):
(WebKit::WebExtensionController::sqliteStore):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.h: Removed.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm: Removed.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp: Added.
(WebKit::rowFilterStringFromRowKeys):
(WebKit::WebExtensionStorageSQLiteStore::WebExtensionStorageSQLiteStore):
(WebKit::WebExtensionStorageSQLiteStore::getAllKeys):
(WebKit::WebExtensionStorageSQLiteStore::getValuesForKeys):
(WebKit::storageSizeOf):
(WebKit::WebExtensionStorageSQLiteStore::getStorageSizeForKeys):
(WebKit::hashmapToVector):
(WebKit::WebExtensionStorageSQLiteStore::getStorageSizeForAllKeys):
(WebKit::WebExtensionStorageSQLiteStore::setKeyedData):
(WebKit::WebExtensionStorageSQLiteStore::deleteValuesForKeys):
(WebKit::WebExtensionStorageSQLiteStore::insertOrUpdateValue):
(WebKit::WebExtensionStorageSQLiteStore::getValuesForAllKeys):
(WebKit::WebExtensionStorageSQLiteStore::getKeysAndValuesFromRowIterator):
(WebKit::WebExtensionStorageSQLiteStore::getAllKeysWithErrorMessage):
(WebKit::WebExtensionStorageSQLiteStore::getValuesForKeysWithErrorMessage):
(WebKit::WebExtensionStorageSQLiteStore::currentDatabaseSchemaVersion):
(WebKit::WebExtensionStorageSQLiteStore::createFreshDatabaseSchema):
(WebKit::WebExtensionStorageSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
(WebKit::WebExtensionStorageSQLiteStore::resetDatabaseSchema):
(WebKit::WebExtensionStorageSQLiteStore::isDatabaseEmpty):
(WebKit::WebExtensionStorageSQLiteStore::databaseURL):
* Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.h: Added.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/289633@main">https://commits.webkit.org/289633@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8875018327f1c4b159dbffafd1320524f6b648ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87454 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6967 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41826 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92316 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38193 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7349 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15132 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67548 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25286 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90456 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5597 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79138 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47890 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5385 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37309 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75803 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34395 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94201 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14619 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10723 "Found 4 new test failures: fast/multicol/simple-line-layout-line-index-after-strut-2.html imported/w3c/web-platform-tests/screen-orientation/onchange-event.html imported/w3c/web-platform-tests/screen-orientation/orientation-reading.html media/video-replaces-poster.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76375 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14873 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74993 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75597 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19972 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18394 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/7557 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13640 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14637 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19932 "Found 1 new failure in UIProcess/Network/NetworkProcessProxy.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14380 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17824 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16163 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -570,13 +573,16 @@ - (void)_didDeleteLocalStorage:(NSNotification *)notification
 
     auto uniqueIdentifier = FileSystem::lastComponentOfPathIgnoringTrailingSlash(storageDirectory);
-    return [[_WKWebExtensionStorageSQLiteStore alloc] initWithUniqueIdentifier:uniqueIdentifier storageType:type directory:storageDirectory usesInMemoryDatabase:NO];
+    return WebExtensionStorageSQLiteStore::create(uniqueIdentifier, type, storageDirectory, false);
```
The `false` param here really should be an `enum class : bool` like `UsesInMemoryDatabase::No` so it is readable inline.

xeenon commented on Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp:
```diff
@@ -184,38 +197,51 @@
-    }];
+    for (auto mapData : oldKeysAndValues) {
+        auto key = mapData.key;
+        auto value = mapData.value;
```
I like `entry` for these iterator vars.
```suggestion
    for (auto entry : oldKeysAndValues) {
        const auto& key = entry.key;
        const auto& value = entry.value;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp:
```diff
@@ -184,38 +197,51 @@
+    // Process new or changed keys
+    for (auto mapData : newKeysAndValues) {
+        auto key = mapData.key;
+        auto value = mapData.value;
```
```suggestion
    for (auto entry : newKeysAndValues) {
        const auto& key = entry.key;
        const auto& value& = entry.value;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp:
```diff
@@ -23,158 +23,171 @@
+            if (keysSuccessfullySet.size() != data.size()) {
+                for (auto i : data.keys()) {
+                    if (!keysSuccessfullySet.contains(i))
+                        data.remove(i);
```
```suggestion
                for (const auto& key : data.keys()) {
                    if (!keysSuccessfullySet.contains(key))
                        data.remove(key);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp:
```diff
@@ -23,158 +23,171 @@
 
-        if (dataType == WebExtensionDataType::Sync && numberOfKeys > webExtensionStorageAreaSyncMaximumItems) {
-            completionHandler(toWebExtensionError(callingAPIName, nullString(), @"exceeded maximum number of items"));
+        if (dataType == WebExtensionDataType::Sync && (size_t)numberOfKeys > webExtensionStorageAreaSyncMaximumItems) {
```
```suggestion
        if (dataType == WebExtensionDataType::Sync && static_cast<size_t>(numberOfKeys) > webExtensionStorageAreaSyncMaximumItems) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp:
```diff
@@ -23,158 +23,171 @@
+    HashMap<String, String> data = { };
+    if (RefPtr json = JSON::Value::parseJSON(dataJSON); json->asObject()) {
+        for (auto i : json->asObject()->keys())
+            data.set(i, json->asObject()->getString(i));
```
```suggestion
        for (const auto& key : json->asObject()->keys())
            data.set(key, json->asObject()->getString(i));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp:
```diff
@@ -23,158 +23,171 @@
+            Ref jsonObject = JSON::Object::create();
+
+            for (auto& i : values)
+                jsonObject->setString(i.key, i.value);
```
```suggestion
            Ref jsonObject = JSON::Object::create();
            for (auto entry : values)
                jsonObject->setString(entry.key, entry.value);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp:
Add Igalia copyright to file.

xeenon commented on Source/WebKit/Sources.txt:
```diff
@@ -572,6 +572,10 @@ UIProcess/Extensions/WebExtensionController.cpp
+UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp
+
+UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp
+
```
```suggestion
```

xeenon reviewed (changes requested):
A couple of API tests to figure out, otherwise looking good with some naming nits.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,428 @@
+    ASSERT(errorMessage.isEmpty());
+
+    RefPtr rows = SQLiteDatabaseFetch(*database(), "SELECT * FROM extension_storage"_s);
+    if (rows)
```
```suggestion
    if (RefPtr rows = SQLiteDatabaseFetch(*database(), "SELECT * FROM extension_storage"_s))
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,428 @@
+    Vector<String> keys;
+    RefPtr rows = SQLiteDatabaseFetch(*database(), "SELECT key FROM extension_storage"_s);
+    if (rows) {
+        RefPtr<WebExtensionSQLiteRow> row = rows->next();
```
```suggestion
    if (RefPtr rows = SQLiteDatabaseFetch(*database(), "SELECT key FROM extension_storage"_s)) {
        RefPtr row = rows->next();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,428 @@
+    ASSERT(errorMessage.isEmpty());
+
+    RefPtr rows = SQLiteDatabaseFetch(*database(), makeString("SELECT * FROM extension_storage WHERE key in ("_s, rowFilterStringFromRowKeys(keys), ")"_s));
+    if (rows)
```
```suggestion
    if (RefPtr rows = SQLiteDatabaseFetch(*database(), makeString("SELECT * FROM extension_storage WHERE key in ("_s, rowFilterStringFromRowKeys(keys), ")"_s)))
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,428 @@
+String WebExtensionStorageSQLiteStore::insertOrUpdateValue(const String& value, const String& key, Ref<WebExtensionSQLiteDatabase> database)
+{
+    assertIsCurrent(queue());
+    DatabaseResult result = SQLiteDatabaseExecute(database, "INSERT OR REPLACE INTO extension_storage (key, value) VALUES (?, ?)"_s, key, value);
```
```suggestion
    assertIsCurrent(queue());

    DatabaseResult result = SQLiteDatabaseExecute(database, "INSERT OR REPLACE INTO extension_storage (key, value) VALUES (?, ?)"_s, key, value);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,428 @@
+
+        Vector<String> keysSuccessfullySet;
+
+        for (auto key : keyedData.keys()) {
```
```suggestion
        Vector<String> keysSuccessfullySet;
        for (auto key : keyedData.keys()) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,428 @@
+    });
+}
+
+static Vector<String> hashmapToVector(HashMap<String, String> map, bool mapKeys)
```
No need to put the param type in the name.
```suggestion
static Vector<String> toVector(HashMap<String, String> map, bool mapKeys)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,428 @@
+    for (auto key : map.keys())
+        totalStorage += key.sizeInBytes();
+
+    for (auto value : map.values())
```
```suggestion
    for (const auto& key : map.keys())
        totalStorage += key.sizeInBytes();

    for (const auto& value : map.values())
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,428 @@
+/*
+ * Copyright (C) 2024 Igalia, S.L. All rights reserved.
```
```suggestion
 * Copyright (C) 2025 Igalia, S.L. All rights reserved.
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.h:
```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2024 Igalia, S.L. All rights reserved.
```
```suggestion
 * Copyright (C) 2025 Igalia, S.L. All rights reserved.
```

xeenon commented on Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2024 Apple Inc. All rights reserved.
+ * Copyright (C) 2024 Igalia S.L. All rights reserved.
```
```suggestion
 * Copyright (C) 2024 Apple Inc. All rights reserved.
 * Copyright (C) 2025 Igalia S.L. All rights reserved.
```

xeenon reviewed (approved):
Looking good.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,425 @@
+            result.append(i);
+    } else {
+        for (auto i : map.values())
+            result.append(i);
```
```suggestion
        for (const auto& key : map.keys())
            result.append(key);
    } else {
        for (const auto& value : map.values())
            result.append(value);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,425 @@
+
+    Vector<String> keys;
+    if (RefPtr rows = SQLiteDatabaseFetch(*database(), "SELECT key FROM extension_storage"_s)) {
+        RefPtr<WebExtensionSQLiteRow> row = rows->next();
```
```suggestion
        RefPtr row = rows->next();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp:
```diff
@@ -0,0 +1,425 @@
+{
+    HashMap<String, String> results;
+
+    RefPtr<WebExtensionSQLiteRow> row = rows->next();
```
```suggestion
    RefPtr row = rows->next();
```

ItsJamie9494 committed:
Port WebExtensionStorageSQLiteStore to C++
https://webkit.org/b/286343

Reviewed by Timothy Hatcher.

This removes the old Cocoa implementation of WebExtensionStorageSQLiteStore, and creates a new c++ storage store
based off the new C++ WebExtensionSQLiteStore

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Extensions/API/WebExtensionContextAPIStorage.cpp: Renamed from Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm.
(WebKit::WebExtensionContext::isStorageMessageAllowed):
(WebKit::WebExtensionContext::storageGet):
(WebKit::WebExtensionContext::storageGetKeys):
(WebKit::WebExtensionContext::storageGetBytesInUse):
(WebKit::WebExtensionContext::storageSet):
(WebKit::WebExtensionContext::storageRemove):
(WebKit::WebExtensionContext::storageClear):
(WebKit::WebExtensionContext::fireStorageChangedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::localStorageStore):
(WebKit::WebExtensionContext::sessionStorageStore):
(WebKit::WebExtensionContext::syncStorageStore):
(WebKit::WebExtensionContext::storageForType):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::getDataRecords):
(WebKit::WebExtensionController::getDataRecord):
(WebKit::WebExtensionController::removeData):
(WebKit::WebExtensionController::calculateStorageSize):
(WebKit::WebExtensionController::removeStorage):
(WebKit::WebExtensionController::sqliteStore):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.h: Removed.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm: Removed.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.cpp: Added.
(WebKit::rowFilterStringFromRowKeys):
(WebKit::WebExtensionStorageSQLiteStore::WebExtensionStorageSQLiteStore):
(WebKit::WebExtensionStorageSQLiteStore::getAllKeys):
(WebKit::WebExtensionStorageSQLiteStore::getValuesForKeys):
(WebKit::storageSizeOf):
(WebKit::WebExtensionStorageSQLiteStore::getStorageSizeForKeys):
(WebKit::hashmapToVector):
(WebKit::WebExtensionStorageSQLiteStore::getStorageSizeForAllKeys):
(WebKit::WebExtensionStorageSQLiteStore::setKeyedData):
(WebKit::WebExtensionStorageSQLiteStore::deleteValuesForKeys):
(WebKit::WebExtensionStorageSQLiteStore::insertOrUpdateValue):
(WebKit::WebExtensionStorageSQLiteStore::getValuesForAllKeys):
(WebKit::WebExtensionStorageSQLiteStore::getKeysAndValuesFromRowIterator):
(WebKit::WebExtensionStorageSQLiteStore::getAllKeysWithErrorMessage):
(WebKit::WebExtensionStorageSQLiteStore::getValuesForKeysWithErrorMessage):
(WebKit::WebExtensionStorageSQLiteStore::currentDatabaseSchemaVersion):
(WebKit::WebExtensionStorageSQLiteStore::createFreshDatabaseSchema):
(WebKit::WebExtensionStorageSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
(WebKit::WebExtensionStorageSQLiteStore::resetDatabaseSchema):
(WebKit::WebExtensionStorageSQLiteStore::isDatabaseEmpty):
(WebKit::WebExtensionStorageSQLiteStore::databaseURL):
* Source/WebKit/UIProcess/Extensions/WebExtensionStorageSQLiteStore.h: Added.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/289633@main
